### PR TITLE
CupertinoTextField: added ability to change placeholder color

### LINKED ISSUE #24175: CupertinoTextField unable to override the default placeholder color.
I'm having problems on overriding the default color for `placeholder` in CupertinoTextField. As I'm trying to set it white it always turns to gray. I've inspected the flutter source base and found this:

https://github.com/flutter/flutter/blob/dfd022956044fe9e336eaa5b93e418befa1918ee/packages/flutter/lib/src/cupertino/text_field.dart#L541-L558

Is there any other way around this?
Here is my flutter installation details:
~~~
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel beta, v0.10.2, on Linux, locale en_US.UTF-8)
[✓] Android toolchain - develop for Android devices (Android SDK 28.0.2)
[✓] Android Studio (version 3.1)
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
[✓] Connected device (1 available)

• No issues found!
~~~

### PR DESCRIPTION
**Pull request by Nizarius**
Resolves #24175 
https://github.com/flutter/flutter/projects/9#card-14775899

### CODE REVIEW

**Comment by Nizarius** [comment]
> I signed it!

**Comment by Nizarius** [comment]
I signed it!

**Comment by Nizarius** [comment]
I signed it

**Comment by Nizarius** [comment]
I signed it

**Comment by Nizarius** [comment]
I signed it

**Comment by Nizarius** [comment]
Should I change anything to resolve this PR?

**Comment by matthew-carroll** [comment]
@Nizarius would you mind correcting the analysis issues?

**Comment by Nizarius** [comment]
@matthew-carroll yeah, of course.
By the way, it seems that it will be better to make not only placeholder color modifiable, but all the placeholder style at once. What do you think? 
If you can make issue for that I can make pr to it. Because it's highly likely that developers want more that change only placeholder color (for example, I want to modify font weight too and so on).

**Comment by matthew-carroll** [comment]
@Nizarius yes, I think we should allow an arbitrary `TextStyle` in a property called `placeholderStyle`.  Let's do that here.  We wouldn't want to merge in color support and then immediately replace it.  Can you update this PR accordingly, and ensure that all checks pass?

Also, would you mind adding some tests for this new behavior?

**Comment by Nizarius** [comment]
@matthew-carroll I updated my PR to pass all the tests.
Frankly speaking, I'm new at Flutter and don't really understand what tests should I add for this new behavior.

**Comment by matthew-carroll** [comment]
Thanks @Nizarius.

@xster let's discuss on Tuesday if we should carry this PR across the finish line.

**Comment by xster** [comment]
Sure. 

@Nizarius, adding tests does sound a bit overwhelming initially though if you take a look at https://github.com/flutter/flutter/blob/master/packages/flutter/test/cupertino/text_field_test.dart, there are already tons of examples that you can base new tests off of. Some even creates and asserts the styles of the placeholder text. I'd create a new test based off of that one to make sure custom placeholder styles are respected.

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 1]
```diff
@@ -149,6 +149,8 @@ class CupertinoTextField extends StatefulWidget {
     this.decoration = _kDefaultRoundedBorderDecoration,
     this.padding = const EdgeInsets.all(6.0),
     this.placeholder,
+    this.placeholderStyle =
+        const TextStyle(fontWeight: FontWeight.w300, color: _kInactiveTextColor),
```
Comment (xster) [inline packages/flutter/lib/src/cupertino/text_field.dart]:
```dart
this.placeholderStyle = const TextStyle(
  fontWeight: FontWeight.w300,
  color: _kInactiveTextColor,
),
```
is our general line break pattern for styling

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 2]
```diff
@@ -177,18 +179,18 @@ class CupertinoTextField extends StatefulWidget {
...
-       assert(maxLengthEnforced != null),
-       assert(scrollPadding != null),
-       assert(maxLines == null || maxLines > 0),
-       assert(maxLength == null || maxLength > 0),
-       assert(clearButtonMode != null),
-       assert(prefixMode != null),
-       assert(suffixMode != null),
-       keyboardType = keyboardType ?? (maxLines == 1 ? TextInputType.text : TextInputType.multiline),
-       super(key: key);
+        assert(autofocus != null),
```
Comment (xster) [inline packages/flutter/lib/src/cupertino/text_field.dart]: Let's not alter these lines if they're unrelated.

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 3]
```diff
@@ -221,6 +223,11 @@ class CupertinoTextField extends StatefulWidget {
   /// main text entry except a lighter font weight and a grey font color.
   final String placeholder;
 
+  /// The style to use for the placeholder text.
+  ///
+  /// Defaults to the [style] property with w300 font weight and grey color.
```
Comment (xster) [inline packages/flutter/lib/src/cupertino/text_field.dart]: Please mention in this doc that it's not just the style that's exclusively used but that it overrides the [style] parameter (which itself inherits from the theme).

Also mention whether it can be null and what does it mean when it is.

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 4]
```diff
@@ -541,9 +549,9 @@ class _CupertinoTextFieldState extends State<CupertinoTextField> with AutomaticK
 
   bool _showSuffixWidget(TextEditingValue text) {
     return widget.suffix != null && _shouldShowAttachment(
-      attachment: widget.suffixMode,
-      hasText: text.text.isNotEmpty,
-    );
+          attachment: widget.suffixMode,
```
Comment (xster) [inline packages/flutter/lib/src/cupertino/text_field.dart]: Leave the existing formatting in unrelated.

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 5]
```diff
@@ -586,17 +595,10 @@ class _CupertinoTextFieldState extends State<CupertinoTextField> with AutomaticK
...
-                maxLines: 1,
-                overflow: TextOverflow.ellipsis,
-                style: textStyle.merge(
-                  const TextStyle(
-                    color: _kInactiveTextColor,
-                    fontWeight: FontWeight.w300,
-                  ),
-                ),
-              ),
+              child: Text(widget.placeholder,
```
Comment (xster) [inline packages/flutter/lib/src/cupertino/text_field.dart]: Follow the existing formatting style. i.e.

```dart
child: Text(
  widget.placeholder,
  maxLines: 1,
  overflow: TextOverflow.ellipsis,
  style: placeholderStyle,
),
```

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 6]
```diff
@@ -606,19 +608,19 @@ class _CupertinoTextFieldState extends State<CupertinoTextField> with AutomaticK
         // First add the explicit suffix if the suffix visibility mode matches.
         if (_showSuffixWidget(text)) {
           rowChildren.add(widget.suffix);
-        // Otherwise, try to show a clear button if its visibility mode matches.
+          // Otherwise, try to show a clear button if its visibility mode matches.
```
Comment (xster) [inline packages/flutter/lib/src/cupertino/text_field.dart]: This seems unrelated too.

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 7]
```diff
@@ -606,19 +608,19 @@ class _CupertinoTextFieldState extends State<CupertinoTextField> with AutomaticK
...
                   : null,
-              child: const Padding(
-                padding: EdgeInsets.symmetric(horizontal: 6.0),
+              child: Padding(
+                padding: const EdgeInsets.symmetric(horizontal: 6.0),
                 child: Icon(
                   CupertinoIcons.clear_thick_circled,
                   size: 18.0,
-                  color: _kInactiveTextColor,
+                  color: placeholderStyle.color,
```
Comment (xster) [inline packages/flutter/lib/src/cupertino/text_field.dart]: I think we're on our own here since you can't change the placeholder's color on iOS itself. Since we don't have a precedent that says these colors are linked, I'd just leave this. (you can always use the suffix widget to put a custom button).

**Comment by Nizarius** [comment]
@xster @matthew-carroll I fixed code according to your notes and also added test for placeholder's style.

**Comment by Nizarius** [comment]
@xster @matthew-carroll hi guys. Is there anything I need to do?

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 8]
```diff
@@ -235,6 +239,13 @@ class CupertinoTextField extends StatefulWidget {
   /// main text entry except a lighter font weight and a grey font color.
   final String placeholder;
 
+  /// The style to use for the placeholder text which overrides [style] property.
+  ///
+  /// If null, placeholder's style will be the same as [style].
```
Comment (matthew-carroll) [inline packages/flutter/lib/src/cupertino/text_field.dart]: This isn't true, is it?  the placeholder should at least have a special color even if `placeholderStyle` isn't provided, right?

**Comment by xster** [comment]
@Nizarius there's some trickiness with the whole TextStyle merge thing that isn't really easy to explain so we tweaked the doc in your branch a bit. Hope that's ok. 

The rest of the PR looks good to us. We'll merge it in when our tree is green. Thanks for the contribution.

**Comment by Nizarius** [comment]
@xster @matthew-carroll thank you, guys.

**Comment by xster** [comment]
Thanks for your contribution
